### PR TITLE
Fix kubectl CRD validation with preserve-unknown-fields objects

### LINKED ISSUE #88252: CRD validation error for x-kubernetes-preserve-unknown-fields: true
**What happened**:
I have CRD with a spec field of type `runtime.RawExtension` and annotation `x-kubernetes-preserve-unknown-fields: true`. I want to store json data in this field which is actually a Grafana Dashboard json.

When I try to create a CR of this type, I am getting validation error from json bytes. `error validating data: unknown object type "nil" in Dashboard.spec.model.gnetId;` . In the json data `.spec.model.gnetId: null`.

**What you expected to happen**:

I expected no error since I set `x-kubernetes-preserve-unknown-fields: true`

**How to reproduce it (as minimally and precisely as possible)**:

```
$ kind create cluster

$ kubectl version --short
Client Version: v1.17.0
Server Version: v1.17.0

# create CRD
$ kubectl create -f https://github.com/searchlight/grafana-operator/raw/f1dc2cce10f9849405d1f45c4fb33fb0221d5e9d/api/crds/grafana.searchlight.dev_dashboards.yaml


# validation error
$ kubectl create -f https://github.com/searchlight/grafana-operator/raw/f1dc2cce10f9849405d1f45c4fb33fb0221d5e9d/hack/examples/dashboard2.yaml

error: error validating "https://github.com/searchlight/grafana-operator/raw/f1dc2cce10f9849405d1f45c4fb33fb0221d5e9d/hack/examples/dashboard2.yaml": error validating data: unknown object type "nil" in Dashboard.spec.model.gnetId; if you choose to ignore these errors, turn validation off with --validate=false


# works, as removed the `.spec.model.gnetId` field
$ kubectl create -f https://github.com/searchlight/grafana-operator/raw/83f8a0fa7da470e79151220ab2b85cae75a38a9c/hack/examples/dashboard2-valid.yaml

dashboard.grafana.searchlight.dev/dashboard2-valid created
```

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): v1.17.0
- Cloud provider or hardware configuration: kind
- OS (e.g: `cat /etc/os-release`): Ubuntu 18.04
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

/sig api-machinery
/area custom-resources

### PR DESCRIPTION
**Pull request by gautierdelorme**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

kubectl doesn't properly handle object fields with `x-kubernetes-preserve-unknown-fields: true` so we have to publish a less precise openapi spec (similar to other pruning we do for the same reason).

**Which issue(s) this PR fixes**:

Fixes https://github.com/kubernetes/kubernetes/issues/88252

**Does this PR introduce a user-facing change?**:

```release-note
Fixed a bug that prevents kubectl to validate CRDs with schema using x-kubernetes-preserve-unknown-fields on object fields.
```

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/v2/conversion.go [hunk 1]
```diff
@@ -81,6 +81,13 @@ func ToStructuralOpenAPIV2(in *structuralschema.Structural) *structuralschema.St
 				changed = true
 			}
 
+			if s.XPreserveUnknownFields && s.Type == "object" {
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/v2/conversion.go:84]: `&& additionalProperties is not set`, right?

Comment (gautierdelorme) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/v2/conversion.go:84]: The bug is still present if `additionalProperties` is set I think.

- You can reproduce it with `additionalProperties` sets to `true`:

```yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  name: foos.k8s.test
spec:
  group: k8s.test
  versions:
    - name: v1
      served: true
      storage: true
      schema:
        openAPIV3Schema:
          type: object
          properties:
            field:
              type: object
              x-kubernetes-preserve-unknown-fields: true
              additionalProperties: true
  scope: Namespaced
  names:
    plural: foos
    singular: foo
    kind: Foo
```

```yaml
apiVersion: k8s.test/v1
kind: Foo
metadata:
  name: foo
field:
  foo: null
```

- You can reproduce it with `additionalProperties` including `nullable: true`:

```yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  name: foos.k8s.test
spec:
  group: k8s.test
  versions:
    - name: v1
      served: true
      storage: true
      schema:
        openAPIV3Schema:
          type: object
          properties:
            field:
              type: object
              x-kubernetes-preserve-unknown-fields: true
              additionalProperties:
                type: object
                nullable: true
  scope: Namespaced
  names:
    plural: foos
    singular: foo
    kind: Foo
```

```yaml
apiVersion: k8s.test/v1
kind: Foo
metadata:
  name: foo
field:
  foo: null
```

Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/v2/conversion.go:84]: ack

**Comment by sttts** [comment]
This deserves backports.

**Comment by gautierdelorme** [comment]
@sttts thanks for the review, I'll work on backports to 1.19 and 1.18.

**Comment by gautierdelorme** [comment]
I'm just trying to unblock the `tide` check which seem to be stuck...
/priority important-soon

**Pull request merged**